File: web/src/lib/career.ts
import { z } from "zod";

export const timeSlotSchema = z.enum(["morning", "afternoon", "evening", "flexible"]);
export const planTaskTypeSchema = z.enum([
  "question_review",
  "structured_expression",
  "business_sense",
  "mock_interview",
  "resume",
  "application",
]);
export const planItemStatusSchema = z.enum(["pending", "in_progress", "completed", "skipped"]);

export const careerProfileSchema = z.object({
  target_role: z.string(),
  target_level: z.string(),
  target_companies: z.array(z.string()),
  preferred_cities: z.array(z.string()),
  weekly_hours: z.number().int().min(1).max(80),
  available_weekdays: z.array(z.number().int().min(0).max(6)).min(1).max(7),
  preferred_time_slot: timeSlotSchema,
  constraints: z.string(),
  confirmed_at: z.string().nullable(),
  updated_at: z.string().nullable(),
});

export const weeklyPlanItemSchema = z.object({
  id: z.string().uuid(),
  plan_id: z.string().uuid().nullable(),
  scheduled_date: z.iso.date(),
  time_slot: timeSlotSchema,
  scheduled_time: z.string().nullable(),
  estimated_minutes: z.number().int().min(5).max(240),
  task_type: planTaskTypeSchema,
  title: z.string().min(1).max(200),
  reason: z.string().min(1).max(600),
  completion_criteria: z.string().min(1).max(500),
  status: planItemStatusSchema,
  origin: z.enum(["ai", "manual", "migrated"]),
  question_id: z.string().uuid().nullable(),
  question_slug: z.string().nullable(),
  coaching_mode: z.enum(["structured_expression", "business_sense"]).nullable(),
  exercise_type: z.string().nullable(),
  difficulty: z.enum(["guided", "assisted", "pressure"]).nullable(),
  position: z.number().int().min(0).max(100),
  completed_at: z.string().nullable(),
});

export const planningBasisSchema = z.object({
  profile_confirmed: z.boolean(),
  question_count: z.number().int().nonnegative(),
  owned_question_count: z.number().int().nonnegative(),
  due_question_count: z.number().int().nonnegative(),
  recent_training_count: z.number().int().nonnegative(),
  evidence_focus: z.string().nullable(),
});

export const weeklyPlanSchema = z.object({
  id: z.string().uuid(),
  week_start: z.iso.date(),
  goal: z.string(),
  items: z.array(weeklyPlanItemSchema),
  status: z.enum(["active", "completed", "archived"]),
  basis: planningBasisSchema,
  model: z.string().nullable(),
  prompt_version: z.string().nullable(),
  skill_version: z.string().nullable(),
  confirmed_at: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const weeklyPlanDraftSchema = z.object({
  id: z.string().uuid(),
  week_start: z.iso.date(),
  goal: z.string(),
  items: z.array(weeklyPlanItemSchema),
  basis: planningBasisSchema,
  model: z.string(),
  prompt_version: z.string(),
  skill_version: z.string(),
  expires_at: z.string(),
});

export const careerQuestionOptionSchema = z.object({
  id: z.string().uuid(),
  slug: z.string(),
  title: z.string(),
  difficulty: z.string(),
  framework: z.string(),
  topics: z.array(z.string()),
  source_document_name: z.string().nullable(),
  review_due: z.boolean(),
  owned: z.boolean(),
});

export const careerWorkspaceSchema = z.object({
  profile: careerProfileSchema,
  weekly_plan: weeklyPlanSchema.nullable(),
  plan_history: z.array(weeklyPlanSchema),
  question_options: z.array(careerQuestionOptionSchema),
});

export const careerProfileConversationResultSchema = z.object({
  reply: z.string().min(1).max(500),
  profile: careerProfileSchema.nullable(),
});

export const careerProfileRequestSchema = z.object({
  target_role: z.string().max(150).default(""),
  target_level: z.string().max(50).default(""),
  target_companies: z.array(z.string().min(1).max(150)).max(20).default([]),
  preferred_cities: z.array(z.string().min(1).max(100)).max(20).default([]),
  weekly_hours: z.number().int().min(1).max(80).default(5),
  available_weekdays: z.array(z.number().int().min(0).max(6)).min(1).max(7),
  preferred_time_slot: timeSlotSchema,
  constraints: z.string().max(2_000).default(""),
}).strict();

export const weeklyPlanDraftRequestSchema = z.object({
  week_start: z.iso.date(),
  instruction: z.string().max(1_000).default(""),
}).strict();
export const careerProfileMessageRequestSchema = z.object({ message: z.string().trim().min(1).max(1_000) }).strict();
export const weeklyPlanRequestSchema = z.object({
  week_start: z.iso.date(),
  goal: z.string().min(1).max(500),
  items: z.array(weeklyPlanItemSchema).min(1).max(20),
  status: z.enum(["active", "completed", "archived"]).default("active"),
  draft_id: z.string().uuid().optional(),
}).strict();
export const weeklyPlanItemStatusRequestSchema = z.object({ status: planItemStatusSchema }).strict();

export type CareerWorkspace = z.infer<typeof careerWorkspaceSchema>;
export type WeeklyPlan = z.infer<typeof weeklyPlanSchema>;
export type WeeklyPlanDraft = z.infer<typeof weeklyPlanDraftSchema>;
export type WeeklyPlanItem = z.infer<typeof weeklyPlanItemSchema>;
export type CareerQuestionOption = z.infer<typeof careerQuestionOptionSchema>;
export type CareerProfileConversationResult = z.infer<typeof careerProfileConversationResultSchema>;

File: api/src/interview_copilot/domain/career.py
from datetime import date, datetime, time
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class CareerProfile(BaseModel):
    model_config = ConfigDict(extra="forbid")

    target_role: str = Field(default="", max_length=150)
    target_level: str = Field(default="", max_length=50)
    target_companies: list[str] = Field(default_factory=list, max_length=20)
    preferred_cities: list[str] = Field(default_factory=list, max_length=20)
    weekly_hours: int = Field(default=5, ge=1, le=80)
    available_weekdays: list[Annotated[int, Field(ge=0, le=6)]] = Field(
        default_factory=lambda: [0, 2, 4, 5], min_length=1, max_length=7
    )
    preferred_time_slot: Literal["morning", "afternoon", "evening", "flexible"] = (
        "evening"
    )
    constraints: str = Field(default="", max_length=2_000)
    confirmed_at: datetime | None = None
    updated_at: datetime | None = None


class WeeklyPlanItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: UUID
    scheduled_date: date
    time_slot: Literal["morning", "afternoon", "evening", "flexible"] = "flexible"
    scheduled_time: time | None = None
    estimated_minutes: int = Field(default=20, ge=5, le=240)
    task_type: Literal[
        "question_review",
        "structured_expression",
        "business_sense",
        "mock_interview",
        "resume",
        "application",
    ]
    title: str = Field(min_length=1, max_length=200)
    reason: str = Field(min_length=1, max_length=600)
    completion_criteria: str = Field(min_length=1, max_length=500)
    status: Literal["pending", "in_progress", "completed", "skipped"] = "pending"
    origin: Literal["ai", "manual", "migrated"] = "manual"
    question_id: UUID | None = None
    question_slug: str | None = Field(default=None, max_length=150)
    coaching_mode: Literal["structured_expression", "business_sense"] | None = None
    exercise_type: str | None = Field(default=None, max_length=40)
    difficulty: Literal["guided", "assisted", "pressure"] | None = None
    position: int = Field(default=0, ge=0, le=100)
    completed_at: datetime | None = None

    @model_validator(mode="after")
    def validate_training_link(self) -> "WeeklyPlanItem":
        if self.task_type in {"structured_expression", "business_sense"} and (
            not self.coaching_mode or not self.exercise_type or not self.difficulty
        ):
            raise ValueError("专项训练事项缺少训练模式、题型或难度")
        if self.question_id and self.task_type not in {
            "question_review",
            "structured_expression",
        }:
            raise ValueError("只有题目学习或结构化表达事项可以关联题目")
        return self


class CareerQuestionOption(BaseModel):
    id: UUID
    slug: str
    title: str
    difficulty: str
    framework: str
    topics: list[str] = Field(default_factory=list)
    source_document_name: str | None = None
    review_due: bool = False
    owned: bool = False


class PlanningBasis(BaseModel):
    profile_confirmed: bool
    question_count: int = Field(ge=0)
    owned_question_count: int = Field(ge=0)
    due_question_count: int = Field(ge=0)
    recent_training_count: int = Field(ge=0)
    evidence_focus: str | None = None


class WeeklyPlanDraft(BaseModel):
    id: UUID
    week_start: date
    goal: str = Field(min_length=1, max_length=500)
    items: list[WeeklyPlanItem] = Field(min_length=1, max_length=20)
    basis: PlanningBasis
    model: str
    prompt_version: str
    skill_version: str
    expires_at: datetime


class WeeklyPlan(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: UUID
    week_start: date
    goal: str = Field(min_length=1, max_length=500)
    items: list[WeeklyPlanItem] = Field(min_length=1, max_length=20)
    status: Literal["active", "completed", "archived"] = "active"
    basis: PlanningBasis
    model: str | None = None
    prompt_version: str | None = None
    skill_version: str | None = None
    confirmed_at: datetime
    created_at: datetime
    updated_at: datetime


class CareerWorkspace(BaseModel):
    profile: CareerProfile
    weekly_plan: WeeklyPlan | None
    plan_history: list[WeeklyPlan] = Field(default_factory=list)
    question_options: list[CareerQuestionOption] = Field(default_factory=list)
